Add Black compatible configurations in documentation (#1366 & #1205) (#1371)
psf/black: docs/compatible_configs.md

Most of the changes that _Black_ make are harmless, but a few do conflict against other
tools. It is not uncommon to be using other tools alongside _Black_ like linters and
type checkers. Some of them need a bit of tweaking to resolve the conflicts. Listed
below are _Black_ compatible configurations for the commonly used tools out there with
their explanations.

## isort

[isort](https://pypi.org/p/isort/) helps to sort and format imports in your Python code.
_Black_ also formats imports, but in a different way from isort's defaults which leads
to conflicting changes.

_Black_ wraps imports that surpass `line-length` by moving identifiers into their own
indented line. If that still doesn't fit the bill, it will put all of them in seperate
lines and put a trailing comma. A more detailed explanation of this behaviour can be
[found here](https://github.com/psf/black#how-black-wraps-lines).

isort should be configured to match _Black_'s import formatting style by these options:

- `multi_line_output = 3`
- `include_trailing_comma = true`
- `force_grid_wrap = 0`
- `combine_as_imports = true`

isort should be configured to wrap imports when they surpass _Black_'s default limit of

> 88 characters via `line_length = 88`.

Author: I am not proud of this. It does a poor job of explaining the configuration. The problem is that I haven't used isort before and I don't know these options well enough to explain them well. Even reading the isort documentation didn't really do much for me.

Author: Well, after realizing my isort configuration was not ideal, I spent some more time reading the isort documentation and finally the options made sense. **Issue resolved.**

# _Black_ compatible configurations

> Most of the changes that _Black_ make are harmless, but a few do conflict against other

Reviewer: All of Black's changes are harmless (or at least, they should be)

Author: Well noted, I think I've been stalking the issue section a bit too much.

parentheses. isort should follow the same behaviour and passing the options
`include_trailing_comma = True` and `use_parentheses = True` configures that.

The option `force_grid_wrap = 0` is just to tell isort to only wrap imports that surpass
the `line_length` limit.

Finally, isort should be told to wrap imports when they surpass _Black_'s default limit
of 88 characters via `line_length = 88`.

### Formats

<details>
<summary>.isort.cfg</summary>

```cfg
[settings]
multi_line_output = 3
include_trailing_comma = True
force_grid_wrap = 0
use_parentheses = True
line_length = 88
```

</details>

<details>
<summary>setup.cfg</summary>

```cfg
[isort]
multi_line_output = 3
include_trailing_comma = True
force_grid_wrap = 0
use_parentheses = True
line_length = 88
```

</details>

<details>
<summary>pyproject.toml</summary>

```toml
[tool.isort]
multi_line_output = 3
include_trailing_comma = True
force_grid_wrap = 0
use_parentheses = True
line_length = 88
```

</details>

<details>
<summary>.editorconfig</summary>

```ini
[*.py]
multi_line_output = 3
include_trailing_comma = True
force_grid_wrap = 0
use_parentheses = True
line_length = 88
```

</details>

> ## flake8

Reviewer 2: ```suggestion
## Flake8
```

> [flake8](https://pypi.org/p/flake8/) is a code linter. It warns you of syntax errors,

Reviewer 2: ```suggestion
[Flake8](https://pypi.org/p/flake8/) is a code linter. It warns you of syntax errors,
```

> possible bugs, stylistic errors, etc. For the most part, flake8 follows

Reviewer 2: ```suggestion
possible bugs, stylistic errors, etc. For the most part, Flake8 follows
```

[PEP 8](https://www.python.org/dev/peps/pep-0008/) when warning about stylistic errors.

> Except, there are a few deviations that cause incompatiblities with _Black_.

Reviewer: ```suggestion
There are a few deviations that cause incompatibilities with _Black_.
```

### Configuration

```
max-line-length = 88
extend-ignore = E203, W503
```

### Why those options above?

When breaking a line, _Black_ will break it before a binary operator. This is compliant
with PEP 8, but this behaviour will cause flake8 to raise
`W503 line break before binary operator` warnings.

In some cases, as determined by PEP 8, _Black_ will enforce an equal amount of

> whitespace around slice operators. Due to this, flake8 will raise

Reviewer 2: ```suggestion
whitespace around slice operators. Due to this, Flake8 will raise
```

`E203 whitespace before ':'` warnings.

> Since both of these warnings are not PEP 8 compliant, flake8 should be configured to

Reviewer 2: ```suggestion
Since both of these warnings are not PEP 8 compliant, Flake8 should be configured to
```

ignore these warnings via `extend-ignore = E203, W503`.

> Also, as like with isort, flake8 should be configured to allow lines up to the length

Reviewer 2: ```suggestion
Also, as like with isort, Flake8 should be configured to allow lines up to the length
```

limit of `88`, _Black_'s default. This explains `max-line-length = 88`.

### Formats

<details>
<summary>.flake8</summary>

```ini
[flake8]
max-line-length = 88
extend-ignore = E203, W503
```

</details>

<details>
<summary>setup.cfg</summary>

```cfg
[flake8]
max-line-length = 88
extend-ignore = E203, W503
```

</details>

<details>
<summary>tox.ini</summary>

```ini
[flake8]
max-line-length = 88
extend-ignore = E203, W503
```

</details>

## Pylint

> [Pylint](https://pypi.org/p/pylint/) is also a code linter like flake8. It has the same

Reviewer 2: ```suggestion
[Pylint](https://pypi.org/p/pylint/) is also a code linter like Flake8. It has the same
```

checks as flake8 and more. In particular, it has more formatting checks regarding style
conventions like variable naming. With so many checks, Pylint is bound to have some
mixed feelings about _Black_'s formatting style.

### Configuration

```
disable = C0330, C0326
max-line-length = 88
```

### Why those options above?

When _Black_ is folding very long expressions, the closing brackets will
[be dedented](https://github.com/psf/black#how-black-wraps-lines).

```py3
ImportantClass.important_method(
    exc, limit, lookup_lines, capture_locals, callback
)
```

> Although, while this style is PEP 8 compliant, Pylint will raise

Reviewer: ```suggestion
Although this style is PEP 8 compliant, Pylint will raise
```
